### PR TITLE
7665 handle the select all case inside the action menu

### LINKED ISSUE #7665: Handle the select all case inside the action menu
## Scope & Context

After the action menu refactoring we don't handle the 'select all' case inside the action menu.

## Technical inputs

Modify the contextStore.
Before we had `contextStoreTargetedRecordIds` which was an array of records.
Now it should look like this:
`contextStoreTargetedRecords`:
```
{
selectedRecords: 'all' | string[],
excludedRecords: string[],
}
```

Modify actions to take this new state into account.

### PR DESCRIPTION
**Pull request by bosiraphael**
Closes #7665 
- Handle select all
- Handle Filters

### CODE REVIEW

File: packages/twenty-front/src/modules/context-store/hooks/useContextStoreSelectedRecords.ts [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+import { useFindManyRecords } from '@/object-record/hooks/useFindManyRecords';
+import { turnFiltersIntoQueryFilter } from '@/object-record/record-filter/utils/turnFiltersIntoQueryFilter';
+import { useFindManyParams } from '@/object-record/record-index/hooks/useLoadRecordIndexTable';
+import { makeAndFilterVariables } from '@/object-record/utils/makeAndFilterVariables';
+import { useRecoilValue } from 'recoil';
+
+const getFilterForSelectedRecords = (
+  selectedRecordIds: string[] | 'all',
+  excludedRecordIds: string[],
+  queryFilter: any,
```
Comment (Weiko) [inline packages/twenty-front/src/modules/context-store/hooks/useContextStoreSelectedRecords.ts:15]:
```suggestion
  queryFilter: RecordGqlOperationFilter | undefined,
): RecordGqlOperationFilter | undefined => {
```

File: packages/twenty-front/src/modules/context-store/hooks/useContextStoreSelectedRecords.ts [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+  );
+
+  const { selectedRecordIds, excludedRecordIds } = contextStoreTargetedRecords;
+
+  // Determine if we should skip the query based on the following conditions:
+  // 1. We're not selecting all records (selectedRecordIds !== 'all')
+  // 2. Either:
+  //    a) No specific records are selected (selectedRecordIds.length === 0)
+  //    b) We're only requesting the 'id' field (which we already have)
+  const isOnlyRequestingId =
```
Comment (Weiko) [inline packages/twenty-front/src/modules/context-store/hooks/useContextStoreSelectedRecords.ts:77]: nit: Usually extracting to a function with a good name can be explicit enough and you won't need a comment. Maybe try extracting isOnlyRequestingId and skip logic in a function
```typescript
  const shouldSkipQuery = (): boolean => {
    const isOnlyRequestingId =
      Object.keys(recordGqlFields ?? {}).length === 1 &&
      recordGqlFields?.id === true;

    return (
      selectedRecordIds !== 'all' &&
      (selectedRecordIds.length === 0 || isOnlyRequestingId)
    );
  };
  
const skip = shouldSkipQuery();
  ````

File: packages/twenty-front/src/modules/context-store/hooks/useContextStoreSelectedRecords.ts [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+
+  const contextStoreCurrentObjectMetadataId = useRecoilValue(
+    contextStoreCurrentObjectMetadataIdState,
+  );
+
+  const contextStoreTargetedRecordsFilters = useRecoilValue(
+    contextStoreTargetedRecordsFiltersState,
+  );
+
+  const { objectMetadataItem } = useObjectMetadataItemById({
```
Comment (Weiko) [inline packages/twenty-front/src/modules/context-store/hooks/useContextStoreSelectedRecords.ts:62]: objectMetadataItem could be undefined, I can see you handle it below in multiple places but don't want want to throw early in this case? Does it make sense to run that code without it?

File: packages/twenty-front/src/modules/object-record/record-index/options/hooks/useRecordData.ts [hunk 4]
```diff
@@ -76,69 +77,63 @@ export const useTableData = ({
...
     totalCount,
     records,
     fetchMoreRecordsWithPagination,
     loading,
   } = useLazyFindManyRecords({
-    ...usedFindManyParams,
+    ...findManyRecordsParams,
+    filter: makeAndFilterVariables([
+      queryFilter,
+      selectedRecordIds !== 'all'
```
Comment (Weiko) [inline packages/twenty-front/src/modules/object-record/record-index/options/hooks/useRecordData.ts:119]: nit: not a big fan of ternary-ception 😅 Up to you but I kind of agree with greptile there

**Review by Weiko — COMMENTED** [review]
@bosiraphael tried locally with more records and delete does not work as expected (only findMany on the first page and does not trigger a delete), I'll let you check!

**Review by charlesBochet — APPROVED** [review]
LGTM!

File: packages/twenty-front/src/modules/context-store/hooks/useContextStoreSelectedRecords.ts (hunk 3 above)
Comment (charlesBochet) [reply packages/twenty-front/src/modules/context-store/hooks/useContextStoreSelectedRecords.ts:63]: +1 we should throw in the useObjectMetadataItemById

File: packages/twenty-front/src/modules/object-record/record-index/options/hooks/useRecordData.ts (hunk 4 above)
Comment (charlesBochet) [reply packages/twenty-front/src/modules/object-record/record-index/options/hooks/useRecordData.ts:119]: +1

File: packages/twenty-front/src/modules/action-menu/actions/record-actions/components/DeleteRecordsActionEffect.tsx [hunk 5]
```diff
@@ -1,51 +1,91 @@
...
+    contextStoreNumberOfSelectedRecordsState,
+  );
+
+  const contextStoreTargetedRecordsRule = useRecoilValue(
+    contextStoreTargetedRecordsRuleState,
+  );
+
+  const filter = computeContextStoreFilters(
+    contextStoreTargetedRecordsRule,
+    objectMetadataItem ?? undefined,
```
Comment (charlesBochet) [inline packages/twenty-front/src/modules/action-menu/actions/record-actions/components/DeleteRecordsActionEffect.tsx:48]: is ?? undefined useful ? Per typing should not be possible

File: packages/twenty-front/src/modules/action-menu/actions/record-actions/components/DeleteRecordsActionEffect.tsx [hunk 6]
```diff
@@ -1,51 +1,91 @@
...
-  const { deleteTableData } = useDeleteTableData({
+  const contextStoreNumberOfSelectedRecords = useRecoilValue(
+    contextStoreNumberOfSelectedRecordsState,
+  );
+
+  const contextStoreTargetedRecordsRule = useRecoilValue(
+    contextStoreTargetedRecordsRuleState,
+  );
+
+  const filter = computeContextStoreFilters(
```
Comment (charlesBochet) [inline packages/twenty-front/src/modules/action-menu/actions/record-actions/components/DeleteRecordsActionEffect.tsx:46]: graphqlFilter

File: packages/twenty-front/src/modules/action-menu/actions/record-actions/components/DeleteRecordsActionEffect.tsx [hunk 7]
```diff
@@ -1,51 +1,91 @@
...
     objectNameSingular: objectMetadataItem?.nameSingular ?? '',
-    recordIndexId: objectMetadataItem?.namePlural ?? '',
+    filter,
   });
 
-  const handleDeleteClick = useCallback(() => {
-    deleteTableData(contextStoreTargetedRecordIds);
-  }, [deleteTableData, contextStoreTargetedRecordIds]);
+  const handleDeleteClick = useCallback(async () => {
+    const recordIdsToDelete = await fetchAllRecordIds();
```
Comment (charlesBochet) [inline packages/twenty-front/src/modules/action-menu/actions/record-actions/components/DeleteRecordsActionEffect.tsx:57]: Ok, we will likely replace this by a more robust deleteMany(filter) implementation later

File: packages/twenty-front/src/modules/action-menu/actions/record-actions/components/RecordActionMenuEntriesSetter.tsx [hunk 8]
```diff
@@ -1,20 +1,38 @@
...
-  if (contextStoreTargetedRecordIds.length === 0) {
+  const contextStoreCurrentObjectMetadataId = useRecoilValue(
+    contextStoreCurrentObjectMetadataIdState,
+  );
+
+  const { objectMetadataItem } = useObjectMetadataItemById({
+    objectId: contextStoreCurrentObjectMetadataId ?? '',
+  });
+
+  if (!objectMetadataItem || !contextStoreNumberOfSelectedRecords) {
```
Comment (charlesBochet) [inline packages/twenty-front/src/modules/action-menu/actions/record-actions/components/RecordActionMenuEntriesSetter.tsx:21]: if !objectMetadataItem ==> we should throw

File: packages/twenty-front/src/modules/context-store/utils/computeContextStoreFilters.ts [hunk 9]
```diff
@@ -0,0 +1,42 @@
+import { ContextStoreTargetedRecordsRule } from '@/context-store/states/contextStoreTargetedRecordsRuleState';
+import { ObjectMetadataItem } from '@/object-metadata/types/ObjectMetadataItem';
+import { RecordGqlOperationFilter } from '@/object-record/graphql/types/RecordGqlOperationFilter';
+import { turnFiltersIntoQueryFilter } from '@/object-record/record-filter/utils/turnFiltersIntoQueryFilter';
+import { makeAndFilterVariables } from '@/object-record/utils/makeAndFilterVariables';
+
+export const computeContextStoreFilters = (
```
Comment (charlesBochet) [inline packages/twenty-front/src/modules/context-store/utils/computeContextStoreFilters.ts:7]: nice! let's add a test on this one

File: packages/twenty-front/src/modules/object-record/record-index/components/RecordIndexContainerContextStoreNumberOfSelectedRecordsEffect.tsx [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+    const { objectMetadataItem } = useObjectMetadataItem({
+      objectNameSingular,
+    });
+
+    const findManyRecordsParams = useFindManyParams(
+      objectMetadataItem?.nameSingular ?? '',
+      objectMetadataItem?.namePlural ?? '',
+    );
+
+    const { totalCount } = useFindManyRecords({
```
Comment (charlesBochet) [inline packages/twenty-front/src/modules/object-record/record-index/components/RecordIndexContainerContextStoreNumberOfSelectedRecordsEffect.tsx:37]: I'm surprise we need

File: packages/twenty-front/src/modules/action-menu/actions/record-actions/components/RecordActionMenuEntriesSetter.tsx (hunk 8 above)
Comment (charlesBochet) [reply packages/twenty-front/src/modules/action-menu/actions/record-actions/components/RecordActionMenuEntriesSetter.tsx:21]: actually, seems that useGetObjectMetadataItemById is throwing and it's not possible anymore to have an undefined objectMetadataItem

**Pull request merged**